File: StriverDP/ninjaTraining.cpp
/*
https://www.codingninjas.com/codestudio/problems/ninja-s-training_3621003?leftPanelTab=1
*/

int ninjaTraining(int n, vector<vector<int>> &points)
{
  vector<int> dp(4,0);
  vector<int> temp(4,0);
  dp[0] = max(points[0][1],points[0][2]);
  dp[1] = max(points[0][0],points[0][2]);
  dp[2] = max(points[0][0],points[0][1]);
  dp[3] = max(points[0][0],max(points[0][1],points[0][2]));
  for(int day = 1 ; day < n ; day++){
    for(int last = 0 ; last < 4 ; last++){
      temp[last] = 0;

      for(int task = 0 ; task<3 ; task++){
        if(task!= last){
          temp[last] = max(temp[last],points[day][task] + dp[task]);
        }
      }
    }
    dp = temp;
  }
  return dp[3];
} 
 

File: StriverDP/ninjaTraining.py
'''
https://www.codingninjas.com/codestudio/problems/ninja-s-training_3621003?leftPanelTab=1
'''
from typing import *

def ninjaTraining(n: int, points: List[List[int]]) -> int:
    dp = [[0]*4 for i in range(n)]
    def td(day:int, task:int):
        maxi = 0
        if(day == 0):
            for i in range(3):
                if(task != i):
                    maxi = max(maxi, points[day][i])
            return maxi

        if(dp[day][task]):
            return dp[day][task]
        point = 0
        for i in range(3):
            if(i != task):
                point = points[day][i] + td(day-1, i)
                maxi = max(maxi, point)
        dp[day][task] = maxi
        return maxi
    
    return td(n-1,3)
